src/components/lib/vmrt/vmrt.c

```diff
@@ -0,0 +1,427 @@
+{
+	vaddr_t page;
+
+	vcpu->shared_mem_cap = capmgr_vm_shared_region_create(&page);
+	vcpu->shared_region = (struct vm_vcpu_shared_region *)page;
+}
+
+void
+vmrt_lapic_access_page_create(struct vmrt_vm_vcpu *vcpu)
+{
+	vcpu->lapic_access_cap = capmgr_vm_lapic_access_create(vcpu->vm->lapic_access_page);
+}
+
+void
+vmrt_vmcb_create(struct vmrt_vm_vcpu *vcpu)
+{
+	vcpu->vmcb_cap = capmgr_vm_vmcb_create(vcpu->vmcs_cap, vcpu->msr_bitmap_cap, vcpu->lapic_access_cap, vcpu->lapic_cap, vcpu->shared_mem_cap, vcpu->handler_tid << 16| vcpu->vpid, 0);
+}
+
+void
+vmrt_vm_vcpu_init(struct vmrt_vm_comp *vm, u32_t vcpu_nr)
+{
+	struct vmrt_vm_vcpu *vcpu;
+	thdid_t tid, handler_tid;
+	thdcap_t cap;
+	vaddr_t resource;
+	assert(vcpu_nr < vm->num_vpu);
+
+	vcpu = &vm->vcpus[vcpu_nr];
+	assert(!vcpu->shared_region);
+
+	vcpu->vm = vm;
+
+	handler_tid = sched_thd_create(vmrt_vm_exception_handler, vcpu);
+	vcpu->handler_tid = handler_tid;
+
+	vcpu->vpid = vcpu_nr + 1;
+	vmrt_vmcs_page_create(vcpu);
+	vmrt_msr_bitmap_page_create(vcpu);
+	vmrt_lapic_page_create(vcpu);
+	vmrt_shared_page_create(vcpu);
+	vmrt_lapic_access_page_create(vcpu);
+	vmrt_vmcb_create(vcpu);
+
+	cap = capmgr_vm_vcpu_create(vcpu->vm->comp_id, vcpu->vmcb_cap, &tid);
+	assert(tid != 0);
+	vcpu->tid = tid;
+	vcpu->cap = cap;
+	vcpu->cpuid = vcpu_nr;
+	vcpu->coreid = cos_coreid();
+
+	vcpu->next_timer = ~0ULL;
+
+}
+
+void
+vmrt_vm_mem_init(struct vmrt_vm_comp *vm, void *mem)
+{
+	vm->guest_addr = mem;
+
+	/* Clean the memory for the VM */
+	memset(mem, 0, vm->guest_mem_sz);
+}
+
+void
+vmrt_vm_create(struct vmrt_vm_comp *vm, char *name, u8_t num_vcpu, u64_t vm_mem_sz)
+{
+	compid_t vm_id = 0;
+	
+	vaddr_t lapic_access_page;
+
+	assert((vm_mem_sz % PAGE_SIZE_4K) == 0);
+
+	vm_id = capmgr_vm_comp_create(vm_mem_sz);
+	vm->comp_id = vm_id;
+
+	/*
+	 * Allocate lapic access page for the VM.
+	 *
+	 * This is to allocate the "lapic access page" for the VM.
+	 * Simply, this is just to allocate a kernel page, mapped it
+	 * into VM comp's EPT at a specified address, by default, it
+	 * is LAPIC_BASE_ADDR. Then, this page needs to be also referenced
+	 * by the VMCS in the kernel.
+	 * 
+	 * This page is only used by a VMCS to let the vcpu know
+	 * it is accessing a lapic page. The real lapic page used by each vcpu
+	 * to read/write data will be allocated later when initializing vcpu.
+	 */
+	lapic_access_page = capmgr_vm_shared_kernel_page_create_at(vm_id, LAPIC_BASE_ADDR);
+	assert(lapic_access_page);
+	vm->lapic_access_page = lapic_access_page;
+
+	assert(num_vcpu < VMRT_VM_MAX_VCPU);
+	vm->num_vpu = num_vcpu;
+	vm->guest_mem_sz = vm_mem_sz;  
+
+	strcpy((char *)&vm->name, name);
+
+	return;
+}
+
+void
+vmrt_vm_data_copy_to(struct vmrt_vm_comp *vm, char *data, u64_t size, paddr_t gpa)
+{
+	char *host_va = GPA2HVA(gpa, vm);
+	memcpy(host_va, data, size);
+}
+
+void
+vmrt_vm_vcpu_resume(struct vmrt_vm_vcpu *vcpu)
+{
+	cos_switch(vcpu->cap, 0, 0, 0, 0, 0);
```

Comment:
I think that if this used the scheduling machinery, this would also set the timer. Not sure how that is done currently. This might be perfectly fine, but isn't the integration as the VMM as a scheduling thread I was expecting.

Reply:
It does not call into the scheduler. This directly calls into the kernel and goes into the vcpu resume. Thus, this directly switches to the vcpu thread. Will need to have more discussion here about the details, but for the initial support, I think this is fine.